### PR TITLE
[percep] Returned distance of closest obstacle

### PR DESCRIPTION
**Pull request by kevinxji**
Program returns the distance of the closest obstacle of ANY PATH (it checks the obstacle distances in `CENTER`, `LEFT`, and `RIGHT` path when `CheckPath` is called, and returns the closest obstacle distance of each of those paths. If `CENTER` has a path, it will return the distance of the closest obstacle from `CENTER`.)

## Compile Test
- [x] Offline Debug: ` ./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=true `
  - [x] No compiler errors or warnings
- [ ] Offline Silent: ` ./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=false `
  - [ ] No compiler errors or warnings
- [x] ZED Debug: ` ./jarvis build jetson/percep -o with_zed=true ar_detection=true obs_detection=true perception_debug=true `
  - [x] No compiler errors, only ZED warnings
- [ ] ZED Silent: ` ./jarvis build jetson/percep -o with_zed=true ar_detection=true obs_detection=true perception_debug=false `
   - [ ] No compiler errors, only ZED warnings
​
## Offline Debug Test
- [x] `./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=true`
- [x] Run code against test1 image set located in perception-files repository
- Check for:
   - [x]  Left and right path are being calculated
   - [x]  Calculated bearing is reasonable given viewer output
   - [x]  AR Tag ID is 2
   - [x]  Obstacles with large width are detected
   - [x]  Path is projected around not through large obstacles
   - [x]  Short obstacles are detected
   - [x]  Steep inclines are detected as obstacles
- [x] Run code against test2 image set located in perception-files repository
- Check for:
   - [x]  Left and right path are being calculated
   - [x]  Calculated bearing is reasonable given viewer output
   - [x]  AR Tag ID is 6
   - [x]  2 AR Tags detected simultaneously
   - [x]  Obstacles with large width are detected
   - [x]  Path is projected around not through large obstacles
   - [x]  Short obstacles are detected
   - [x]  Steep inclines are detected as obstacles
- [x] Run code against distance_test image set located in perception-files repository
   - [x] Distance values returned are 11ft, 8ft, and 5ft
   - [x] Margin of error within 1 foot
- [x] Run code against distance_test1 image set located in perception-files repository
   - [x] Distance values returned are 12ft, 9ft, 5ft, and 3ft
   - [x] Margin of error within 1 foot
​
## Offline Silent Test
- [ ] `./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=false`
- [ ] Run `time ./jarvis exec jetson/percep` against test_case1 image set
- [ ] Run `time ./jarvis exec jetson/percep` against test_case2 image set
- [ ] User value is less than or equal to \<TODO Determine Value\>
​
## Unit Test
- [x] ` ./jarvis build jetson/percep -o with_zed=true ar_detection=true obs_detection=true perception_debug=true `
- [x] `./jarvis exec lcm_tools_echo Obstacle "/obstacle"`
- [x] `./jarvis exec lcm_tools_echo TargetList "/target_list"`
- Test environment contains:
  - [x] 1 large obstacle ~ over 40cm tall
  - [x] 2 small obstacles ~  25cm-30cm tall
  - [x] 1 post
  - [x] 1 hill
- Test for:
  - [x] Detected large obstacle 6m away
  - [x] Detected small obstacle 2m away
  - [x] Detected clear path when 1.2m gap between 1 large and 1 small obstacle
  - [x] Left and right path are being calculated
  - [x] Calculated bearing is reasonable given viewer output of path
  - [x] AR Tag ID is the correct ID
  - [x] Obstacles with large width are detected
  - [x] Short obstacles are detected
  - [ ] FPS values don't exceed \<TODO find this value in Issue #552\>
  - [x] Obstacle.lcm in /obstacle is populated correctly
  - [x] TargetList.lcm in /target_list is populated correctly
  - [x] Tested with tape measure at various distances (Acceptable error of less than 0.5 meters)
  - [x] Tested FP filtering
​
## Integration Test
- [ ] Create new branch based off of main
- [ ] Checkout new branch
- [ ] Cherry-Pick all PRs that need testing onto new branch
- [ ] Compile all Software Packages
- [ ] Run all code simultaneously
- [ ] Obstacle.lcm in /obstacle is populated correctly
- [ ] TargetList.lcm in /target_list is populated correctly
- [ ] Sensors are Active
  - [ ] Stereo Camera (ZED)
  - [ ] Inertial Measurement Unit (IMU)
  - [ ] Global Positioning System (GPS)

### CODE REVIEW

File: jetson/percep/main.cpp [hunk 1]
```diff
@@ -146,11 +146,9 @@ int main() {
...
      //Update LCM 
     obstacleMessage.bearing = lastObstacle.bearing; //update LCM bearing field
-    if(lastObstacle.distance <= obstacle_detection.distance)
-      obstacleMessage.distance = (lastObstacle.distance/1000); //update LCM distance field
-    else
-      obstacleMessage.distance = (obstacle_detection.distance/1000); //update LCM distance field
-    cerr << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!Path Sent: " << obstacleMessage.bearing << "\n";
+    obstacleMessage.distance = lastObstacle.distance; //update LCM distance field
+    cerr << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!Path Bearing Sent: " << obstacleMessage.bearing << "\n";
+    cerr << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!Distance Sent: " << obstacleMessage.distance << "\n";
```
Comment (DavidSmith166) [inline jetson/percep/main.cpp:151]: Is there a reason for the formatting of this error message? It seems excessive

File: jetson/percep/pcl.cpp [hunk 2]
```diff
@@ -157,11 +157,11 @@ void PCL::FindInterestPoints(std::vector<pcl::PointIndices> &cluster_indices,
     for (int i = 0; i < (int)cluster_indices.size(); ++i)
     {
         std::vector<int>* curr_cluster = &interest_points[i];
-
+        
         //Initialize interest points
         std::fill(curr_cluster->begin(), curr_cluster->end(), cluster_indices[i].indices[0]);
         
-        //Order of interest points: 0=Up Left 1=Up Right 2=Low Right 3=Low Left
+
```
Comment (DavidSmith166) [inline jetson/percep/pcl.cpp:164]: Has the order changed or is this just removing documentation?

File: jetson/percep/main.cpp (hunk 1 above)
Comment (mmartin4972) [reply jetson/percep/main.cpp:151]: Since we currently have a lot of output it's very helpful to have the offset for debugging purposes. We are currently working on cleaning up our debug output in Issue #460

File: jetson/percep/pcl.cpp [hunk 3]
```diff
@@ -236,26 +237,44 @@ double PCL::FindClearPath(const std::vector<std::vector<int>> &interest_points,
     
     //Check Center Path
     if(CheckPath(interest_points, viewer, obstacles, compareLine(0,-HALF_ROVER), compareLine(0,HALF_ROVER))) {
-        std::cout << "CENTER PATH IS CLEAR!!!" << std::endl;
+            std::cerr << "CENTER PATH IS CLEAR!!!" << std::endl;
+            std::cerr << "Distance from Last Obstacle: " << distance << " meters" << std::endl;
```
Comment (mmartin4972) [inline jetson/percep/pcl.cpp:241]: We print this out in main, so no need here

File: jetson/percep/pcl.cpp [hunk 4]
```diff
@@ -236,26 +237,44 @@ double PCL::FindClearPath(const std::vector<std::vector<int>> &interest_points,
...
+        double centerDistance, leftDistance, rightDistance;
+        centerDistance = distance; //from previous loop of CheckPath, where it was only looking at the center path for a clear path
+
+        //Initialize base cases outside of scope
+        vector<int> centerObstacles = {obstacles.at(0), obstacles.at(1)};
 
-    //Initialize base cases outside of scope
-    vector<int> centerObstacles = {obstacles.at(0), obstacles.at(1)};
+        //Find Clear left path
+        std::cerr << "Checking Left Path" << std::endl;
```
Comment (mmartin4972) [inline jetson/percep/pcl.cpp:254]: Can you change cerrs to couts?

Comment (mmartin4972) [reply jetson/percep/pcl.cpp:254]: Just a consistency thing we're trying to do. Only change them in the ones in this function

File: jetson/percep/pcl.cpp [hunk 5]
```diff
@@ -236,26 +237,44 @@ double PCL::FindClearPath(const std::vector<std::vector<int>> &interest_points,
...
-    //Find clear right path
-    double rightAngle = getAngleOffCenter(10, 1, interest_points, viewer, obstacles);
+        //return smallest distance of an obstacle from all the paths
+        if(rightDistance < leftDistance && rightDistance < centerDistance) distance = rightDistance/1000.0;
+        else if(leftDistance < rightDistance && leftDistance < centerDistance) distance = leftDistance/1000.0;
+        else distance = centerDistance/1000.0;
 
-    //Return the smallest angle (left if equal)
-    return fabs(rightAngle) < fabs(leftAngle) ? rightAngle : leftAngle;
+        std::cerr << "Distance from Last Obstacle: " << distance << " meters" << std::endl;
```
Comment (mmartin4972) [inline jetson/percep/pcl.cpp:273]: Same as above

Comment (mmartin4972) [reply jetson/percep/pcl.cpp:273]: Probly don't need

File: jetson/percep/pcl.hpp [hunk 6]
```diff
@@ -3,6 +3,8 @@
 
 #include "perception.hpp"
 #include <pcl/common/common_headers.h>
+#include <math.h>
+#include <float.h>
```
Comment (mmartin4972) [inline jetson/percep/pcl.hpp:7]: Why were these added?

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

File: jetson/percep/pcl.cpp (hunk 2 above)
Comment (kevinxji) [reply jetson/percep/pcl.cpp:164]: I don't believe it has, probably was an accident. Will recommit the commit

File: jetson/percep/pcl.hpp (hunk 6 above)
Comment (kevinxji) [reply jetson/percep/pcl.hpp:7]: In order to use `DBL_MAX`, I needed the `float.h` library. `math.h` was before when I used `pow()`. Will get rid of `math.h`

**Comment by mmartin4972** [comment]
Silent mode currently not functioning for this PR. Will be solved and testable in PR #562

**Review by mmartin4972 — DISMISSED** [review]

**Review by DavidSmith166 — APPROVED** [review]

**Comment by mmartin4972** [comment]
I think you had some merge conflicts that you resolved on this branch. Can you pull those conflicts onto your computer and perform test2 again to make sure everything is still working fine after you resolved those conflicts?

File: jetson/percep/main.cpp (hunk 1 above)
Comment (arschallwig) [reply jetson/percep/main.cpp:151]: Are we leaving cerrs as is? Or should we start moving towards couts?

File: jetson/percep/pcl.cpp [hunk 7]
```diff
@@ -280,26 +280,41 @@ double PCL::FindClearPath(const std::vector<std::vector<int>> &interest_points,
...
     if(CheckPath(interest_points, viewer, obstacles, compareLine(0,-HALF_ROVER), compareLine(0,HALF_ROVER))) {
-        std::cout << "CENTER PATH IS CLEAR!!!" << std::endl;
+            std::cout << "CENTER PATH IS CLEAR!!!" << std::endl;
         return 0;
     }
+    else { //otherwise, check to find left and right paths
+
+        //Values that store the distances of the last obstacle from a given CheckPath
+        double centerDistance, leftDistance, rightDistance;
+        centerDistance = distance; //from previous loop of CheckPath, where it was only looking at the center path for a clear path
```
Comment (arschallwig) [inline jetson/percep/pcl.cpp:290]: May want to consider initializing centerDistance in the same line as the others are declared (289) just for efficiency and clarity

File: jetson/percep/pcl.cpp [hunk 8]
```diff
@@ -316,8 +331,15 @@ bool PCL::CheckPath(const std::vector<std::vector<int>> &interest_points,
 
     bool end = true;
     
+    double previousDistance = DBL_MAX;
+
+    //if there are no interest points, the distance from the last obstacle should be DBL_MAX
+    distance = previousDistance;
```
Comment (arschallwig) [inline jetson/percep/pcl.cpp:337]: If DBL_MAX is a constant, is there a reason its not just assigned to distance in 337, rendering 334 unnecessary? And if it is not a constant, maybe the naming convention should be clarified

Comment (kevinxji) [reply jetson/percep/pcl.cpp:337]: `distance` is ultimately the variable that is returned. `previousDistance` and `currentDistance` are compared to each other to see whether `previousDistance` is less than the `currentDistance`, since we want to return the smallest distance. Technically, `distance` could be just `currentDistance`, but I used `currentDistance` to make it more clear (also `distance` is a variable in PCL)

Comment (kevinxji) [reply jetson/percep/pcl.cpp:337]: DBL_MAX is a constant value from the <float.h> class (from C++)

**Review by mmartin4972 — APPROVED** [review]

**Review by arschallwig — APPROVED** [review]
All the changes were addressed, looks good.

**Review by DavidSmith166 — APPROVED** [review]

File: jetson/percep/main.cpp (hunk 1 above)
Comment (mmartin4972) [reply jetson/percep/main.cpp:151]: we are moving towards cout. Don't have to worry ab fixing that in this PR
